freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UsersStateDifference.java

```diff
@@ -5,113 +5,187 @@
-
-        Multimap<String, String> groupMembershipToRemove = HashMultimap.create();
-        ipaGroupMembership.forEach((group, user) -> {
-            if (!umsGroupMembership.containsEntry(group, user)) {
-                LOGGER.debug("removing user : {} to group : {}", user, group);
-                groupMembershipToRemove.put(group, user);
-            }
-        });
-
-        LOGGER.info("groupMembershipToRemove size= {}", groupMembershipToRemove.size());
-
-        Set<String> usersToRemove =
-            getUsersToBeRemoved(
-                umsState.getGroupMembership().get(UserServiceConstants.CDP_USERSYNC_INTERNAL_GROUP),
-                ipaState.getGroupMembership().get(UserServiceConstants.CDP_USERSYNC_INTERNAL_GROUP));
-        LOGGER.info("usersToRemove size= {}", usersToRemove.size());
+    public static UsersStateDifference fromUmsAndIpaUsersStates(UmsUsersState umsState, UsersState ipaState) {
+        ImmutableSet<FmsUser> usersToAdd = calculateUsersToAdd(umsState, ipaState);
+        LOGGER.info("usersToAdd size = {}", usersToAdd.size());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("adding users {}", usersToAdd.stream().map(FmsUser::getName).collect(Collectors.toSet()));
+        }
+
+        ImmutableSet<String> usersToRemove = calculateUsersToRemove(umsState, ipaState);
+        LOGGER.info("usersToRemove size = {}", usersToRemove.size());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("removing users {}", usersToRemove.stream().collect(Collectors.toSet()));
+        }
+
+        ImmutableSet<FmsGroup> groupsToAdd = calculateGroupsToAdd(umsState, ipaState);
+        LOGGER.info("groupsToAdd size = {}", groupsToAdd.size());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("adding groups {}", groupsToAdd.stream().map(FmsGroup::getName).collect(Collectors.toSet()));
+        }
+
+        ImmutableSet<FmsGroup> groupsToRemove = calculateGroupsToRemove(umsState, ipaState);
+        LOGGER.info("groupsToRemove size = {}", groupsToRemove.size());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("removing groups {}", groupsToRemove.stream().map(FmsGroup::getName).collect(Collectors.toSet()));
+        }
+
+        ImmutableMultimap<String, String> groupMembershipToAdd = calculateGroupMembershipToAdd(umsState, ipaState);
+        LOGGER.info("groupMembershipToAdd size = {}", groupMembershipToAdd.size());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("adding group memberships {}", groupMembershipToAdd.asMap());
+        }
+
+        ImmutableMultimap<String, String> groupMembershipToRemove = calculateGroupMembershipToRemove(umsState, ipaState);
+        LOGGER.info("groupMembershipToRemove size = {}", groupMembershipToRemove.size());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("removing group memberships {}", groupMembershipToRemove.asMap());
```

Comment:
overall things lgtm, but I think this log could be moved into the calculate methods.
also `isDebugEnabled` is overkill I think, as we have debug enabled everywhere, and if we are planning to change that then we will have to revisit our logging anyway

Reply:
done